### PR TITLE
add uninitializedSlice

### PR DESCRIPTION
**Pull request by 9il**

### CODE REVIEW

File: std/experimental/ndslice/slice.d [hunk 1]
```diff
@@ -611,6 +611,43 @@ pure nothrow unittest
 }
 
 /++
+Creates an untinitilized array and an n-dimensional slice over it.
```
Comment (andralex) [inline std/experimental/ndslice/slice.d]: What is "untinitilized"? Is that a typo that propagated through the whole doc?

File: std/experimental/ndslice/slice.d [hunk 2]
```diff
@@ -611,6 +611,43 @@ pure nothrow unittest
...
+Params:
+    lengths = list of lengths for each dimension
+    slice = slice to copy shape and data from
+Returns:
+    untinitilized n-dimensional slice
++/
+Slice!(Lengths.length, Select!(replaceArrayWithPointer, T*, T[]))
+untinitilizedSlice(T,
+    Flag!`replaceArrayWithPointer` replaceArrayWithPointer = Yes.replaceArrayWithPointer,
```
Comment (andralex) [inline std/experimental/ndslice/slice.d]: Just use `Flag!"replaceArrayWithPointer" f = Yes.replaceArrayWithPointer`. After using `Flag` for a while I figured the name of the flag itself does not matter at all; its meaning is already present in the type and the constants involved.

File: std/experimental/ndslice/slice.d [hunk 3]
```diff
@@ -732,6 +766,66 @@ auto makeSlice(T,
...
+    alloc = allocator
+    lengths = list of lengths for each dimension
+    init = default value for array initialization
+    slice = slice to copy shape and data from
+Returns:
+    a structure with fields `array` and `slice`
++/
+SliceAllocationResult!(Lengths.length, T, replaceArrayWithPointer)
+makeUninitializedSlice(T,
+    Flag!`replaceArrayWithPointer` replaceArrayWithPointer = Yes.replaceArrayWithPointer,
```
Comment (andralex) [inline std/experimental/ndslice/slice.d]: ditto

File: std/experimental/ndslice/slice.d [hunk 4]
```diff
@@ -732,6 +766,66 @@ auto makeSlice(T,
...
+    Allocator,
+    Lengths...)(auto ref Allocator alloc, Lengths lengths)
+    if (allSatisfy!(isIndex, Lengths) && Lengths.length)
+{
+    return .makeUninitializedSlice!(T, replaceArrayWithPointer, Allocator)(alloc, [lengths]);
```
Comment (andralex) [inline std/experimental/ndslice/slice.d]: This allocates an array at each call. How about:

```d
typeof(Lengths[0])[Lengths.length] t = [ Lengths ];
return .makeUninitializedSlice!(T, replaceArrayWithPointer, Allocator)(alloc, t[]);
```

File: std/experimental/ndslice/slice.d [hunk 5]
```diff
@@ -732,6 +766,66 @@ auto makeSlice(T,
...
+}
+
+/// ditto
+SliceAllocationResult!(N, T, replaceArrayWithPointer)
+makeUninitializedSlice(T,
+    Flag!`replaceArrayWithPointer` replaceArrayWithPointer = Yes.replaceArrayWithPointer,
```
Comment (andralex) [inline std/experimental/ndslice/slice.d]: ditto

File: std/experimental/ndslice/slice.d (hunk 1 above)
Comment (9il) [inline std/experimental/ndslice/slice.d]: Oh, yes

File: std/experimental/ndslice/slice.d (hunk 4 above)
Comment (9il) [inline std/experimental/ndslice/slice.d]: No, DMD FE is pretty clever here:

```d
size_t bar(size_t[2] ab) @nogc
{
	return ab[0] + ab[1];
}

size_t foo(size_t a, int b) @nogc
{
	return bar([a, b]);
}
```
Compiles

**Comment by 9il** [comment]
Rebased:
1. Type fixed
2. Flags name fixed for whole module
3. No, DMD FE is pretty clever here:
```d
size_t bar(size_t[2] ab) @nogc
{
    return ab[0] + ab[1];
}

size_t foo(size_t a, int b) @nogc
{
    return bar([a, b]);
}

```

For some reason github does not remove msgs for fixed lines anymore.

@andralex Thank you for the review!

**Comment by JackStouffer** [comment]
Just like `iotaSlice`, I really fail to see the utility of this when it only saves you a couple of lines.

```
import std.array;

auto a = uninitializedArray!(float[])(30_000);
auto slice = a.sliced(30, 10, 100);
```

vs

```
auto slice = uninitializedSlice!float(30, 10, 100);
```

**Comment by andralex** [comment]
@JackStouffer makes a solid point, especially considering the function is niche in the first place. @9il?

**Comment by 9il** [comment]
1. iotaSlice is faster then sliced iota. This is important for tensor generation.

2. uninitializedSlice was not added before because I had the same thought. However, after I got enough experience I found that:
 - ndslice variant is more safe, because dimensions are runtime values. Segmentation fault errors are big problem.
 - scientists can use ndslice without other modules. Furthermore, the do not need to understand D ranges and D arrays to use ndslices. They just need stdio and `for` loop.
 - It is very weird to write this 2 lines 8 times in the same function. I expect that ndslice variant will be used more frequently because numeric problems are more adopted to uninitialized data.

**Comment by andralex** [comment]
>iotaSlice is faster then sliced iota. This is important for tensor generation.

Why?

>uninitializedSlice was not added before because I had the same thought. However, after I got enough experience I found that:

The point remains that although this PR adds almost 100 lines, the payload of it (code that actually does work) is focused in four functions. Two are forwarding one-liners (which wouldn't be needed in the first place if we didn't provide the facility), one is a 2-liner (which I'd write as a 1-liner), and one is a 1-liner. Looks like the cards don't stack well.

So the question here is whether `uninitializedSlice` is something that users need often enough to provide it as a first-class facility. If not, let's make the code an example and move on.

**Comment by 9il** [comment]
> So the question here is whether uninitializedSlice is something that users need often enough to provide it as a first-class facility. If not, let's make the code an example and move on.

Yes, users need it often enough. As I said I expect that ndslice variant will be used more frequently.
Simple example is `gemm`:
```d
auto c = uninitialisedSlice!double!(a.length!0, b.length!1);
for(...)
{
    ...
    gemm(ctx, 1.0, a, b, 0.0, c); // c := a * b
}
```

Both Mir (except few cases) and ndslice do not allocate data in algorithms. All data should be allocated y user. So `uninitialisedSlice` is usual method.

**Comment by andralex** [comment]
Auto-merge toggled on

**Comment by 9il** [comment]
Thanks!

**Pull request merged**